pkg/schemacmp/table.go

```diff
@@ -235,6 +235,168 @@ func (indexMap) JoinWithNil(value Lattice) (Lattice, error) {
+		if len(av) != len(bv) {
+			return false
+		}
+		for j := 0; j < len(av); j++ {
+			if av[j] != bv[j] {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// encodePartitionDefinitionToLattice collects the necessary information for comparing a partition.
+func encodePartitionDefinitionToLattice(pd model.PartitionDefinition) Lattice {
+	return EqualitySingleton(partitionDefinition{pd.LessThan, pd.InValues})
+}
+
+type partitionMap map[string]Lattice
+
+func (partitionMap) New() LatticeMap {
+	return make(partitionMap)
+}
+
+func (a partitionMap) Get(key string) Lattice {
+	val, ok := a[key]
+	if !ok {
+		return nil
+	}
+	return val
+}
+
+func (a partitionMap) Insert(key string, value Lattice) {
+	a[key] = value
+}
+
+func (a partitionMap) ForEach(f func(key string, value Lattice) error) error {
+	for key, value := range a {
+		if err := f(key, value); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (partitionMap) CompareWithNil(value Lattice) (int, error) {
+	return 1, nil
+}
+
+func (partitionMap) ShouldDeleteIncompatibleJoin() bool {
+	return false
+}
+
+func (partitionMap) JoinWithNil(value Lattice) (Lattice, error) {
+	return value, nil
+}
+
+const (
+	partitionTupleIndexType = iota
+	partitionTupleIndexExpr
+	partitionTupleIndexColumns
+	partitionTupleIndexEnable
+	partitionTupleIndexDefinitions
+	partitionTupleIndexNum
+)
+
+func encodePartitionInfoToLattice(pi *model.PartitionInfo) Tuple {
+	partitions := make(partitionMap, len(pi.Definitions))
+	for _, d := range pi.Definitions {
+		partitions[d.Name.L] = encodePartitionDefinitionToLattice(d.Clone())
+	}
+	columns := make(Tuple, 0, len(pi.Columns))
+	for _, c := range pi.Columns {
+		columns = append(columns, Singleton(c.L))
+	}
+	num := Singleton(pi.Num)
+	// Only range and list partition num can be changed.
+	if pi.Type == model.PartitionTypeRange || pi.Type == model.PartitionTypeList {
+		num = Uint64(pi.Num)
+	}
+	return Tuple{
+		Singleton(pi.Type),
+		Singleton(pi.Expr),
+		columns,
+		Singleton(pi.Enable),
+		Map(partitions),
+		num,
+	}
+}
+
+func restorePartitionInfoFromUnwrapped(ctx *format.RestoreCtx, info []interface{}) {
+	ctx.WriteKeyWord(" PARTITION BY ")
+	typ := info[partitionTupleIndexType].(model.PartitionType)
+	ctx.WriteKeyWord(typ.String())
+
+	if expr := info[partitionTupleIndexExpr].(string); len(expr) > 0 {
+		ctx.WritePlainf("(%s)", expr)
+	} else {
+		ctx.WriteKeyWord(" COLUMNS")
+		ctx.WritePlain("(")
+		for i, colName := range info[partitionTupleIndexColumns].([]interface{}) {
+			if i != 0 {
+				ctx.WritePlain(", ")
+			}
+			ctx.WritePlain(colName.(string))
+		}
+		ctx.WritePlain(")")
+	}
+	if typ != model.PartitionTypeHash {
+		ctx.WritePlain(" (")
+		for i, pair := range sortedMap(info[partitionTupleIndexDefinitions].(map[string]interface{})) {
+			if i > 0 {
+				ctx.WritePlain(", ")
+			}
+			ctx.WriteKeyWord("PARTITION ")
+			ctx.WriteName(pair.key)
+
+			d := pair.value.(partitionDefinition)
+			if len(d.lessThan) > 0 {
```

Comment:
in create table AST, value less than could be an expression. So here did it evaluate and only save the result?

Reply:
yes, table info only saved the string results.